internal/exporter/pulsarexporter/pulsar_exporter.go

```diff
@@ -17,91 +17,94 @@ package pulsarexporter
 import (
 	"context"
 	"fmt"
-	"log"
-	"time"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
 
 	"github.com/apache/pulsar-client-go/pulsar"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/pdata/pmetric"
-	"go.uber.org/zap"
 )
 
 var errUnrecognizedEncoding = fmt.Errorf("unrecognized encoding")
 
-// pulsarMetricsProducer uses sarama to produce metrics messages to pulsar
-type pulsarMetricsProducer struct {
+// pulsarMetricsExporter produce metrics messages to pulsar
+type pulsarMetricsExporter struct {
+	client    pulsar.Client
 	producer  pulsar.Producer
+	topic     string
 	marshaler MetricsMarshaler
 	logger    *zap.Logger
 }
 
-type pulsarErrors struct {
-	err string
-}
-
-func (pe pulsarErrors) Error() string {
-	return fmt.Sprintf("Failed to deliver messages due to %s", pe.err)
-}
-
-func (e *pulsarMetricsProducer) metricsDataPusher(_ context.Context, md pmetric.Metrics) error {
+func (e *pulsarMetricsExporter) metricsDataPusher(ctx context.Context, md pmetric.Metrics) error {
 	messages, err := e.marshaler.Marshal(md)
 	if err != nil {
 		return consumererror.NewPermanent(err)
 	}
-	for _, element := range messages { // (1) pulsar at the time of implementing this did not support sending an array of messages. (2) Use array index if needed to propagate back the errors
-		_, err = e.producer.Send(context.Background(), element)
+	var errors error
+	for _, element := range messages {
+		e.producer.SendAsync(ctx, element, func(_ pulsar.MessageID, _ *pulsar.ProducerMessage, err error) {
+			if err != nil {
+				errors = multierr.Append(errors, err)
+			}
+		})
 	}
-	if err != nil {
-		log.Fatal(err)
-		return pulsarErrors{err.Error()}
+	if errors == nil{
+		return nil
 	}
+	return fmt.Errorf("pulsar producer failed to send metric data due to error: %w", errors)
+}
+
+func (e *pulsarMetricsExporter) Close(context.Context) error {
+	e.producer.Close()
 	return nil
 }
 
-func newPulsarProducer(config Config) (pulsar.Producer, error) {
+func newPulsarProducer(config Config) (pulsar.Client, pulsar.Producer, error) {
```

Comment:
Does this need to return a client and a producer? It doesn't seem like the client is used.